### ISSUE TITLE
Type mismatch after migration

### ISSUE THREAD

**Issue by HPIWicket** [realm/realm-java#846]
In our migration we delete a field and afterwards we get in some cases a "type mismatch. Was BOOLEAN, expected INTEGER" for the field isUploaded

Our class looks like: 

```
public class PictureFile extends RealmObject {
    private int id;
    private String fileName;
    // private String description;
    private boolean isUploaded;

    // Getter and Setter for all fields
}
```

This happened to us in with multiple tables. We could see that the row indices caches in the static field columnIndices of the Realm class don't correspond to the actual indices of the table (checked with getTable for debugging).

For the migration we use the following code:

```
Table pictureFileTable = realm.getTable(PictureFile.class);
if (getIndexForProperty(pictureFileTable, "description") >= 0) {
    pictureFileTable.removeColumn(getIndexForProperty(pictureFileTable, "description"));
}
```

We have multiple migrations on other tables before, but that is the first column removal on this table.

How can we fix the migration to avoid this problem, because restarting doesn't fix the problem and app data has to be cleared. 

We also observed the inconsistency between the Realm.columnIndices field the actual file when we added a column in the migration. This is because the migration always adds it to the end, but we didn't but the new field after all the existing field. Therefore the indices were messed up. For example:

```
public static class Test extends RealmObject {
       private int oldField;
       private int oldField2;
       private int newField; // added by migration
       private int oldField3;
}
```

The indices for newField and oldField3 are messed up when the migration was executed, otherwise everything was fine.

Any help is appreciated

Greeting
Wicket

**Comment by cmelchior** [comment]
Hi @HPIWicket 
Thank you for a detailed bug report and sorry for the late response. I have tried to reproduce the behaviour you are describing but cannot. Can you look at this PR and see what the difference is to your code?

https://github.com/realm/realm-java/pull/848/files

**Comment by HPIWicket** [comment]
Hey,

Thanks for the answer. We only encountered the problem when querying the Realm afterwards. The migration finished successfully and we could acquire a Realm instance, but the query failed.

**Comment by cmelchior** [comment]
Hi @HPIWicket 
I have tried to reproduce the error after your updated description, and I cannot see any failures. Also by looking at the code, I *think* it should work correctly.

- What version of Realm are you using?
- What query are you trying to execute?
- Can you provide a small sample app that inhibits this behaviour?

**Comment by cmelchior** [comment]
Closing due to no response and we cannot reproduce it

**Comment by HPIWicket** [comment]
Hey,

sorry for the late answer. I tried to reproduce it with a sample app, without success. We still get reportings of this error in out app. If I can write an application to reproduce the error I will post again.

Thanks for the help :)

**Comment by kneth** [comment]
@cmelchior Can you reapply your old pull request so we can resolve it?

**Comment by cmelchior** [comment]
Hi @HPIWicket 
I think I have found the error. It basically happens because you are using 2 realms where one is updated using RealmMigration and another is created from fresh AND the migration code adds fields in another order than in the model class. Previously we cached field indices globally which triggered this error when querying the Realm.

Realms are not corrupted however, so this PR should work for any existing apps that experience this problem
